src/ngraph_assign_clusters.cc

```diff
@@ -689,50 +700,88 @@ Status AssignClusters(Graph* graph) {
       bool both_src_dst_are_encapsulates =
           src_encapsulate >= 0 && dst_encapsulate >= 0;
       bool src_dst_are_distinct = src_encapsulate != dst_encapsulate;
-      if (both_src_dst_are_encapsulates && src_dst_are_distinct) {
-        // vector of reasons why 2 clusters were separated
-        vector<string> reasons_for_this_separation;
-        for (auto& inner_itr : it.second) {
-          reason_count[inner_itr]++;
-          reasons_for_this_separation.push_back(reason_string[inner_itr]);
-        }
-        auto reasons_string = ng::join(reasons_for_this_separation);
-        if (reasons_string.find(reason_string[0]) != std::string::npos) {
-          return errors::Internal(
-              "UNSUPPORTED should not be a reason why 2 encapsulates did not "
-              "merge, because unsupported ops would not end up in "
-              "encapsulates",
-              reasons_string);
-        }
-        std::unordered_map<string, int> pair_reason_hist;
-        for (auto& i : ng::split(reasons_string, ',')) {
-          pair_reason_hist[i]++;
+      vector<int> reason_count_encapsulates_for_pair(num_reasons, 0);
+      bool pair_has_reason = false;
+      string deadness_string = "";
+      for (auto& inner_itr : it.second) {
+        // This if checks if the pair are 2 distinct encapsulates
+        // In which case it asserts certain non-merging reasons are not possible
+        // And also prints the reasons of non-merging
+        if (both_src_dst_are_encapsulates && src_dst_are_distinct) {
+          if (is_forbidden_reason(inner_itr)) {
+            return errors::Internal(
+                inner_itr,
```

Comment:
Just a question: this will eliminate the all-zero summary of two non-merged encapsulates, right?

Reply:
because of this check thee logs will print only 4 viable reasons for encapsulate connecting edges:
`0->1: DEADNESS:1, BACKEND:0, STATICINPUT:0, PATHEXISTS:0`

Note that the 4 viable reasons could still be all zeros. It just filters out the not-possible reasons